dotnet/src/AutoGen/Core/Middleware/FunctionCallMiddleware.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// FunctionCallMiddleware.cs
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.AI.OpenAI;
+
+namespace AutoGen.Core.Middleware;
+
+/// <summary>
+/// The middleware that process function call message that both send to an agent or reply from an agent.
+/// </summary>
+public class FunctionCallMiddleware : IMiddleware
+{
+    private readonly IEnumerable<FunctionDefinition>? functions;
+    private readonly IDictionary<string, Func<string, Task<string>>>? functionMap;
+    public FunctionCallMiddleware(
+        IEnumerable<FunctionDefinition>? functions = null,
+        IDictionary<string, Func<string, Task<string>>>? functionMap = null,
+        string? name = null)
+    {
+        this.Name = name ?? nameof(FunctionCallMiddleware);
+        this.functions = functions;
+        this.functionMap = functionMap;
+    }
+
+    public string? Name { get; }
+
+    public async Task<Message> InvokeAsync(MiddlewareContext context, IAgent agent, CancellationToken cancellationToken = default)
+    {
+        // if the last message is a function call message, invoke the function and return the result instead of sending to the agent.
+        var lastMessage = context.Messages.Last();
+        if (lastMessage is not null && lastMessage is { Content: null, FunctionName: string functionName, FunctionArguments: string functionArguments })
+        {
+            if (this.functionMap?.TryGetValue(functionName, out var func) is true)
+            {
+                var result = await func(functionArguments);
+                return new Message(role: Role.Function, content: result, from: agent.Name)
+                {
+                    FunctionName = functionName,
+                    FunctionArguments = functionArguments,
+                };
+            }
+            else if (this.functionMap is not null)
+            {
+                var errorMessage = $"Function {functionName} is not available. Available functions are: {string.Join(", ", this.functionMap.Select(f => f.Key))}";
+
+                return new Message(role: Role.Function, content: errorMessage, from: agent.Name)
+                {
+                    FunctionName = functionName,
+                    FunctionArguments = functionArguments,
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException("FunctionMap is not available");
+            }
+        }
+
+        // combine functions
+        var options = new GenerateReplyOptions(context.Options ?? new GenerateReplyOptions());
+        var combinedFunctions = this.functions?.Concat(options.Functions ?? []) ?? options.Functions;
+        options.Functions = combinedFunctions?.ToArray();
+
+        var reply = await agent.GenerateReplyAsync(context.Messages, options, cancellationToken);
```

Comment:
This place gets a little interesting. So each middleware is a wrapper around an inner agent, which gets called inside the `InvokeAsync` function? 

If that's the case, why do we process the reply below and try to execute the function?

Reply:
There're several strategies that can be applied here
- invoke immediately and return result only: this strategy makes the function_call invisible to other agents and makes the most compatibility
- invoke immediately and return [function_call, function_call_result], similar to have user proxy agent execute function, but only require one agent to complete
- return function_call object only: the common two agent chat pattern

The current implementation is if `function_map` is provided, the first strategy will be applied, and if not, the third strategy will be applied.

Reply:
I see. So, in case of (1), does the function call result get fed back to the LLM middleware so it can produce something based on the result? 

What is an example use case of (2)?

Reply:
@ekzhu 
Not for case (1). For case (1) the result from function call will be used as answer. the case(3) is for fedding function result back to LLM.

For use of case (2), it will be similar to case(3) but will rely on another agent to produce something based on the result.

Reply:
I see. Though this makes it not exactly like other agents like Langchain and ChatGPT, as tool calls for them are internal conversations.

Reply:
For Langchain and ChatGPT I believe the strategy they use is case (1) (tool call is private to other agents). For our case, we can provide multiple options for developers.

Reply:
So if user wants to have the langchain / chatgpt experience, how does case (1) work? From the change it seems that the function call result is returned but the LLM doesn't get a chance to take another look at the function call result and make one more inference to say something about the result.

Reply:
Oh I might mis-understand here.So the langchain/chatgpt experience would be case (4) then. Which workflow will be like below

```
while (reply is function_call)
{
   function_result = invoke_function()
   reply = agent.GenerateReplyAsync([...chatHistory, reply, function_result])
}

return reply
```